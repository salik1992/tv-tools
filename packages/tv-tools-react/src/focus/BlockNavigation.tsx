import type { PropsWithChildren } from 'react';
import { FocusContext, useFocusContainer } from '../focus';

function markAsProcessed() {
	return true;
}

/**
 * BlockNavigation is a component that prevents navigation to escape from its children.
 * Useful for Modal, VirtualKeyboard, etc.
 */
export const BlockNavigation = ({ children }: PropsWithChildren) => {
	const { focusContextValue, useOnLeft, useOnUp, useOnDown, useOnRight } =
		useFocusContainer();

	useOnLeft(markAsProcessed);
	useOnRight(markAsProcessed);
	useOnUp(markAsProcessed);
	useOnDown(markAsProcessed);

	return (
		<FocusContext.Provider value={focusContextValue}>
			{children}
		</FocusContext.Provider>
	);
};
