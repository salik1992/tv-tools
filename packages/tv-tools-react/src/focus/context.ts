import { createContext } from 'react';
import { noop } from '@salik1992/tv-tools/utils/noop';
import type { Focus } from './types';

export const FocusContext = createContext<Focus>({
	addChild: noop,
});
