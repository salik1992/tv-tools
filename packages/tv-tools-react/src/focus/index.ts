export { FocusRoot } from './FocusRoot';
export { HorizontalFocus } from './HorizontalFocus';
export { Interactable } from './Interactable';
export { VerticalFocus } from './VerticalFocus';
export { useFocusContainer } from './useFocusContainer';
export { useKeySequenceListener } from './useKeySequenceListener';
