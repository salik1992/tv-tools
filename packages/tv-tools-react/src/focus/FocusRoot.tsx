import type { FocusEvent, KeyboardEvent, PropsWithChildren } from 'react';
import { useCallback } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { focus } from '@salik1992/tv-tools/focus';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager.
 * Use at the root of your application.
 */
export const FocusRoot = ({
	children,
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents,
}: PropsWithChildren<{
	alwaysPreventNavigationalEvents?: boolean;
}>) => {
	const onFocus = useCallback((event: FocusEvent) => {
		focus.handleFocusEvent(event.nativeEvent);
	}, []);

	const keyDownCapture = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keydown', 'capture', event);
	}, []);

	const keyDownBubble = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keydown', 'bubble', event);
		if (
			alwaysPreventNavigationalEvents &&
			isDirectional(event) &&
			document.activeElement?.tagName !== 'INPUT'
		) {
			event.preventDefault();
			event.stopPropagation();
		}
	}, []);

	const keyUpCapture = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keyup', 'capture', event);
	}, []);

	const keyUpBubble = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keyup', 'bubble', event);
		if (
			alwaysPreventNavigationalEvents &&
			isDirectional(event) &&
			document.activeElement?.tagName !== 'INPUT'
		) {
			event.preventDefault();
			event.stopPropagation();
		}
	}, []);

	return (
		<div
			id="focusRoot"
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			{children}
		</div>
	);
};
