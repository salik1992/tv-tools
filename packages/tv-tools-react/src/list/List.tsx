import {
	type KeyboardEvent,
	type WheelEvent,
	type ReactNode,
	useMemo,
	useState,
	useEffect,
} from 'react';
import type {
	ListImplementation,
	RenderDataElement,
} from '@salik1992/tv-tools/list';
import { FocusContext, useFocusContainer } from '../focus';
import { useThrottledCallback } from '../utils/useThrottledCallback';

export const List = <
	Implementation extends ListImplementation,
	Configuration extends
		ConstructorParameters<Implementation>[1] = ConstructorParameters<Implementation>[1],
>({
	id,
	Implementation,
	configuration,
	renderItem,
	orientation = 'horizontal',
	focusOnMount = false,
	throttleMs = 300,
}: {
	id?: string;
	Implementation: Implementation;
	configuration: Omit<Configuration, 'id'>;
	renderItem: (element: RenderDataElement) => ReactNode;
	orientation?: 'vertical' | 'horizontal';
	focusOnMount?: boolean;
	throttleMs?: number;
}) => {
	const {
		container,
		focusContextValue,
		useOnUp,
		useOnDown,
		useOnLeft,
		useOnRight,
	} = useFocusContainer(id);
	const list = useMemo(
		() =>
			new Implementation(container, {
				id: container.id,
				...configuration,
			}),
		[Implementation, configuration, container],
	);
	const [renderData, setRenderData] = useState(list.getRenderData());

	useEffect(() => {
		if (focusOnMount) {
			container.focus({ preventScroll: true });
		}
	}, [focusOnMount]);

	const backward = useThrottledCallback(
		(e: KeyboardEvent<HTMLElement>) => {
			const newRenderData = list.moveBy(-1, (e.target as HTMLElement).id);
			if (newRenderData === renderData) {
				return false;
			}
			setRenderData(newRenderData);
			return true;
		},
		[list, renderData],
		{ throttledReturn: false, limitMs: throttleMs },
	);

	const forward = useThrottledCallback(
		(e: KeyboardEvent<HTMLElement>) => {
			const newRenderData = list.moveBy(1, (e.target as HTMLElement).id);
			if (newRenderData === renderData) {
				return false;
			}
			setRenderData(newRenderData);
			return true;
		},
		[list, renderData],
		{ throttledReturn: false, limitMs: throttleMs },
	);

	if (orientation === 'horizontal') {
		// @ts-ignore: TODO reversed limitation
		useOnLeft(backward, [backward]);
		// @ts-ignore: TODO reversed limitation
		useOnRight(forward, [forward]);
	} else {
		// @ts-ignore: TODO reversed limitation
		useOnUp(backward, [backward]);
		// @ts-ignore: TODO reversed limitation
		useOnDown(forward, [forward]);
	}

	const onWheel = useThrottledCallback(
		(e: WheelEvent<HTMLElement>) => {
			const delta = e.deltaX || e.deltaY;
			const targetId = (e.target as HTMLElement).id;
			let newRenderData = renderData;
			if (delta < 0) {
				newRenderData = list.moveBy(-1, targetId);
			} else if (delta > 0) {
				newRenderData = list.moveBy(1, targetId);
			}
			if (newRenderData !== renderData) {
				setRenderData(newRenderData);
				e.preventDefault();
				e.stopPropagation();
			}
		},
		[list, renderData],
		{ throttledReturn: undefined, limitMs: throttleMs },
	);

	return (
		<FocusContext.Provider value={focusContextValue}>
			<div className="list" onWheel={onWheel}>
				<div
					className="list-inner-wrap"
					style={{
						transform: `translate${orientation === 'horizontal' ? 'X' : 'Y'}(-${renderData.listOffset}px)`,
					}}
				>
					{renderData.elements.map(renderItem)}
				</div>
			</div>
		</FocusContext.Provider>
	);
};
