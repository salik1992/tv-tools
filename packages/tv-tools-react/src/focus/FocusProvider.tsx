import type { ComponentProps, FocusEvent, KeyboardEvent } from 'react';
import { useCallback, useMemo } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { FocusManager } from '@salik1992/tv-tools/focus';
import { noop } from '@salik1992/tv-tools/utils/noop';
import { FocusContext } from './context';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager and
 * it also creates the focus manager instance for the application which is available through
 * the FocusContext or useFocusManager hook.
 * Use at the root of your application.
 * @prop alwaysPreventNavigationalEvents - Use this to prevent TV use their own navigation.
 * @prop ...props - Any other props that you want to pass to the wrapping div element.
 */
export const FocusProvider = ({
	children,
	alwaysPreventNavigationalEvents,
	onFocus: onOuterFocus,
	onKeyDownCapture: onOuterKeyDownCapture,
	onKeyDown: onOuterKeyDown,
	onKeyUpCapture: onOuterKeyUpCapture,
	onKeyUp: onOuterKeyUp,
	...props
}: ComponentProps<'div'> & {
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents?: boolean;
}) => {
	const focusManager = useMemo(() => new FocusManager(), []);

	const onFocus = useCallback(
		(event: FocusEvent<HTMLDivElement>) => {
			focusManager.handleFocusEvent(event.nativeEvent);
			onOuterFocus?.(event);
		},
		[focusManager, onOuterFocus],
	);

	const keyDownCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'capture', event);
			onOuterKeyDownCapture?.(event);
		},
		[focusManager, onOuterKeyDownCapture],
	);

	const keyDownBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyDown?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyDown],
	);

	const keyUpCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'capture', event);
			onOuterKeyUpCapture?.(event);
		},
		[focusManager, onOuterKeyUpCapture],
	);

	const keyUpBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyUp?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyUp],
	);

	const focusContextValue = useMemo(
		() => ({
			focusManager,
			addChild: noop,
		}),
		[focusManager],
	);

	return (
		<div
			{...props}
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			<FocusContext.Provider value={focusContextValue}>
				{children}
			</FocusContext.Provider>
		</div>
	);
};
