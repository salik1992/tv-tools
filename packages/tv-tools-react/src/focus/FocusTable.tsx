import {
	type ComponentProps,
	type ComponentType,
	type PropsWithChildren,
	type ReactNode,
	createContext,
	useCallback,
	useContext,
	useEffect,
	useMemo,
} from 'react';
import type { TableFocusContainer } from '@salik1992/tv-tools/focus';
import { FocusContext } from './context';
import { useTableFocusContainer } from './useTableFocusContainer';

const ChildrenOnlyComponent = ({ children }: PropsWithChildren) => {
	return <>{children}</>;
};

const TableContext = createContext<{
	container?: TableFocusContainer;
	TrComponent: ComponentType<PropsWithChildren<unknown>>;
	TdComponent: ComponentType<PropsWithChildren<unknown>>;
}>({
	TrComponent: ChildrenOnlyComponent,
	TdComponent: ChildrenOnlyComponent,
});

const FocusTd = ({
	rowSpan,
	colSpan,
	children,
	...props
}: PropsWithChildren<{
	colSpan?: number;
	rowSpan?: number;
}>) => {
	const { addChild, ...originalfocusContextValue } = useContext(FocusContext);
	const { TdComponent } = useContext(TableContext);

	const addChildWithSpans = useCallback(
		(childId: string) => {
			addChild(childId, { rowSpan, colSpan });
		},
		[addChild, rowSpan, colSpan],
	);

	const focusContextValue = useMemo(
		() => ({ ...originalfocusContextValue, addChild: addChildWithSpans }),
		[addChildWithSpans, originalfocusContextValue],
	);

	return (
		<FocusContext.Provider value={focusContextValue}>
			<TdComponent {...props}>{children}</TdComponent>
		</FocusContext.Provider>
	);
};

const FocusTr = (props: PropsWithChildren<unknown>) => {
	const { TrComponent, container } = useContext(TableContext);

	useEffect(() => {
		// Needs to run every time to maintain children order.
		container?.addRow();
	});

	return <TrComponent {...props} />;
};

export type FocusTableRenderComponents<
	TrComponent extends ComponentType,
	TdComponent extends ComponentType,
> = {
	FocusTr: (props: ComponentProps<TrComponent>) => ReactNode;
	FocusTd: (
		props: {
			colSpan?: number;
			rowSpan?: number;
		} & ComponentProps<TdComponent>,
	) => ReactNode;
};

export const FocusTable = ({
	id,
	children,
	// @ts-expect-error: TODO
	TableComponent = ChildrenOnlyComponent,
	TrComponent = ChildrenOnlyComponent,
	TdComponent = ChildrenOnlyComponent,
	...tableProps
}: {
	children: (
		components: FocusTableRenderComponents<
			typeof TrComponent,
			typeof TdComponent
		>,
		container: TableFocusContainer,
	) => ReactNode;
	id?: string;
	TableComponent?: ComponentType;
	TrComponent?: ComponentType;
	TdComponent?: ComponentType;
} & ComponentProps<typeof TableComponent>) => {
	const {
		focusContextValue,
		container,
		useOnLeft,
		useOnUp,
		useOnDown,
		useOnRight,
	} = useTableFocusContainer(id);

	useOnLeft(
		(event) =>
			container.moveFocus({ x: -1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnRight(
		(event) =>
			container.moveFocus({ x: 1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnUp(
		(event) =>
			container.moveFocus({ y: -1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnDown(
		(event) =>
			container.moveFocus({ y: 1 }, (event.target as HTMLElement).id),
		[container],
	);

	const tableContextValue = useMemo(
		() => ({
			container,
			TrComponent,
			TdComponent,
		}),
		[TrComponent, TdComponent],
	);

	return (
		<TableContext.Provider value={tableContextValue}>
			<FocusContext.Provider value={focusContextValue}>
				<TableComponent {...tableProps}>
					{children({ FocusTr, FocusTd }, container)}
				</TableComponent>
			</FocusContext.Provider>
		</TableContext.Provider>
	);
};
