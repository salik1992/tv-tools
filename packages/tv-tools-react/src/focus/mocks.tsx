import type { RefObject } from 'react';
import { useContext } from 'react';
import type { FocusManager } from '@salik1992/tv-tools/focus';
import { FocusContext } from './context';

export const ExposeFocusManager = ({
	focusManager,
}: {
	focusManager: RefObject<FocusManager | null>;
}) => {
	const context = useContext(FocusContext);
	if (context.focusManager) {
		focusManager.current = context.focusManager;
	}
	return null;
};

export function assertFocusManager(
	focusManager: RefObject<FocusManager | null>,
): asserts focusManager is RefObject<FocusManager> {
	if (focusManager.current === null) {
		throw new Error('FocusManager is null');
	}
}
