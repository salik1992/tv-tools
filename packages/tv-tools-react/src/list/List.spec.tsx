import { type PropsWithChildren, createRef } from 'react';
import { act, render } from '@testing-library/react';
import { LEFT, RIGHT } from '@salik1992/tv-tools/control';
import type { FocusManager } from '@salik1992/tv-tools/focus';
import { BasicList } from '@salik1992/tv-tools/list/BasicList';
import { Performance } from '@salik1992/tv-tools/utils/Performance';
import { FocusProvider, Interactable } from '../focus';
import { ExposeFocusManager, assertFocusManager } from '../focus/mocks';
import { List } from './List';

const throttleMs = 300;
jest.useFakeTimers();

const focusManager = createRef<FocusManager>();
const Wrapper = ({ children }: PropsWithChildren) => (
	<FocusProvider id="focusRoot">
		{children}
		<ExposeFocusManager focusManager={focusManager} />
	</FocusProvider>
);

describe('List', () => {
	it('should render the list', () => {
		const { container, rerender, unmount } = render(null, {
			wrapper: Wrapper,
		});

		assertFocusManager(focusManager);
		const focusSpy = jest.spyOn(focusManager.current, 'focus');

		rerender(
			<List
				id="list"
				Implementation={BasicList}
				throttleMs={throttleMs}
				configuration={{
					performance: Performance.BASIC,
					dataLength: 5,
					visibleElements: 5,
					config: {
						navigatableElements: 3,
						scrolling: {
							first: 50,
							other: 100,
						},
					},
				}}
				renderItem={({ id, dataIndex, onFocus, offset }) => (
					<Interactable
						id={id}
						key={id}
						data-data-index={dataIndex}
						data-offset={offset}
						onFocus={onFocus}
						onPress={jest.fn()}
					/>
				)}
				focusOnMount
			/>,
		);
		expect(container.innerHTML).toMatchSnapshot();
		expect(focusSpy).toHaveBeenCalledWith('list-0', {
			preventScroll: true,
		});
		unmount();
	});

	it('should move with keyboard, mouse wheel and virtual arrows', () => {
		const { container, rerender, unmount } = render(null, {
			wrapper: Wrapper,
		});

		assertFocusManager(focusManager);
		const focusSpy = jest.spyOn(focusManager.current, 'focus');

		rerender(
			<List
				id="list"
				Implementation={BasicList}
				throttleMs={throttleMs}
				configuration={{
					performance: Performance.BASIC,
					dataLength: 10,
					visibleElements: 7,
					config: {
						navigatableElements: 5,
						scrolling: {
							first: 50,
							other: 100,
						},
					},
				}}
				renderItem={({ id, dataIndex, onFocus, offset }) => (
					<Interactable
						id={id}
						key={id}
						data-data-index={dataIndex}
						data-offset={offset}
						onFocus={onFocus}
						onPress={jest.fn()}
					/>
				)}
				focusOnMount
			/>,
		);

		// RIGHT
		act(() =>
			container
				.querySelector('#list-0')
				?.dispatchEvent(RIGHT.toKeyboardEvent('keydown')),
		);
		expect(focusSpy).toHaveBeenNthCalledWith(2, 'list-1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// Wheel forward
		act(() =>
			container
				.querySelector('#list-1')
				?.dispatchEvent(
					new WheelEvent('wheel', { deltaY: 1, bubbles: true }),
				),
		);
		expect(focusSpy).toHaveBeenNthCalledWith(3, 'list-2', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// Next arrow
		act(() =>
			container
				.querySelector('.mouse-arrow.next')
				?.dispatchEvent(new MouseEvent('click', { bubbles: true })),
		);
		expect(focusSpy).toHaveBeenNthCalledWith(4, 'list-3', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// Previous arrow
		act(() =>
			container
				.querySelector('.mouse-arrow.previous')
				?.dispatchEvent(new MouseEvent('click', { bubbles: true })),
		);
		expect(focusSpy).toHaveBeenNthCalledWith(5, 'list-2', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// Wheel backward
		act(() =>
			container
				.querySelector('#list-2')
				?.dispatchEvent(
					new WheelEvent('wheel', { deltaY: -1, bubbles: true }),
				),
		);
		expect(focusSpy).toHaveBeenNthCalledWith(6, 'list-1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// LEFT
		act(() =>
			container
				.querySelector('#list-1')
				?.dispatchEvent(LEFT.toKeyboardEvent('keydown')),
		);
		expect(focusSpy).toHaveBeenNthCalledWith(7, 'list-0', {
			preventScroll: true,
		});

		unmount();
	});
});
