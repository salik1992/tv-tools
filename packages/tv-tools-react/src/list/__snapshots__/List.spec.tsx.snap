// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render the list 1`] = `"<div id="focusRoot"><div class="list"><div class="list-inner-wrap" style="transform: translateX(-0px);"><div data-data-index="0" data-offset="0" id="list-0" tabindex="0"></div><div data-data-index="1" data-offset="0" id="list-1" tabindex="0"></div><div data-data-index="2" data-offset="0" id="list-2" tabindex="0"></div><div data-data-index="3" data-offset="0" id="list-3" tabindex="0"></div><div data-data-index="4" data-offset="0" id="list-4" tabindex="0"></div></div><div class="mouse-arrow next"></div></div></div>"`;
